NUP-2306 Delayed Link serialization

vitaly-krugl opened the pull request:
Fixes #1198.

Second PR for NUP-2306 that adds capnproto serialization support for the delayed link feature. The first (merged) PR added delayed link functionality without serialization.

Per offline discussion with @scottpurdy, Delayed Link serialization support is only added to the capnproto serialization logic (but not in the yaml-based bundle serialization logic)

vitaly-krugl commented:
@scottpurdy, I implemented serialization and deserialization of the delayed link info using capnp. It passes all the existing nupic.core tests (with default to delay=0). I am now working on a non-default test. Please take a look at c6f0d363d7d113ad8f3dc59f3a8f3e1b1bfed64d for concept. We can factor out the array copy logic into a separate file, but for now I just want to get it working.

scottpurdy commented on src/nupic/engine/Link.cpp:
```diff
@@ -388,6 +388,17 @@ void Link::purgeBufferHead()
 
+
+template <typename ST, typename ABT>
+void copyArrayToArrayProto(ST src, ABT builder, size_t elementCount)
```
Would be nice to put this outside `Link.cpp` fro reuse since other classes will probably use `ArrayProto`

scottpurdy commented on src/nupic/engine/Link.cpp:
```diff
@@ -396,17 +407,164 @@ void Link::write(LinkProto::Builder& proto) const
+
+    const auto arrayType = array.getType();
+
+    switch (arrayType)
```
This switch should go entirely inside `copyArrayToArrayProto`, right?

vitaly-krugl replied:
@scottpurdy: No, because `copyArrayToArrayProto` is a templated function that relies on the source buffer type and builder type to be pre-determined before the call to 'copyArrayToArrayProto`.

scottpurdy replied:
`copyArrayToArrayProto` won't need to be templated anymore since we can determine appropriate types inside the switch/case blocks. If you don't want to repeat the for loop inside every `case` then you can have a generic `copyArrayToArrayProto` with the switch statement and then a simple `_tempaltedCopyArrayToArrayProto` like the current one.

The main point is that we don't want the big `switch` statement in the client code - we want it in the utility function for arrays that can be reused in different places.

scottpurdy commented on src/nupic/engine/Link.cpp:
```diff
@@ -396,17 +407,164 @@ void Link::write(LinkProto::Builder& proto) const
 
+
+template <typename DDT, typename ART>
+void copyArrayProtoToArray(ART reader, Array & dest,
```
Keep this right next to `copyArrayToArrayProto`

vitaly-krugl replied:
ok

scottpurdy commented on src/nupic/engine/Link.cpp:
```diff
@@ -396,17 +407,164 @@ void Link::write(LinkProto::Builder& proto) const
+
+    auto unionSelection = genericArrayReader.which();
+
+    switch (unionSelection)
```
Similarly, move this switch into `copyArrayProtoToArray` if possible

vitaly-krugl replied:
@scottpurdy: not possible; similar reason as for `copyArrayToArrayProto`

vitaly-krugl commented:
@scottpurdy: I responded to your review feedback. Also, the latest commit contains the delayed link serialization test.

However, something isn't right. During de-serialization, [`Link::read`](https://github.com/numenta/nupic.core/blob/f8f7cfe6e97dd3d2293fc638f483bc178911239c/src/nupic/engine/Link.cpp#L512) isn't getting called. See [test]( https://github.com/numenta/nupic.core/blob/f8f7cfe6e97dd3d2293fc638f483bc178911239c/src/test/unit/engine/InputTest.cpp#L417). Subsequently, the test either fails with an unexpected value or segmentation fault.

Perhaps, the test is instrumenting serialization/de-serialization incorrectly. I looked at different serialization examples in nupic.core unit tests, and there is some variation between them. I based my serialization on the example `experimental/ExtendedTemporalMemoryTest.cpp`.

vitaly-krugl commented:
@scottpurdy, referring to my previous comment,  I think I found the reason why `Link::read` isn't getting called: [`Network::read`](https://github.com/numenta/nupic.core/blob/f8f7cfe6e97dd3d2293fc638f483bc178911239c/src/nupic/engine/Network.cpp#L1069-L1070) appears to contain its own version of the link de-serialization logic. It's seems ugly that this exists alongside `Link::read`.

vitaly-krugl commented:
@scottpurdy, I moved the switch statements out of Link::read/Link::write into utility functions. Will move the utility functions to separate file in a later phase. FYI

vitaly-krugl commented on src/nupic/engine/Network.cpp:
```diff
@@ -363,7 +363,9 @@ Network::link(const std::string& srcRegionName, const std::string& destRegionNam
-  destInput->addLink(linkType, linkParams, srcOutput, propagationDelay);
+  auto link = new Link(linkType, linkParams, srcOutput, destInput,
+                       propagationDelay);
+  destInput->addLink(link, srcOutput);
```
This change and the one below are minor code cleanup that take advantage of the new separation of logic for instantiating a Link versus adding the Link to the Input.

vitaly-krugl commented:
@scottpurdy, the nupic.core tests pass now. On Thursday, I will move the Array/ArrayProto copy logic into separate utility files and also make sure that all Nupic tests pass.

scottpurdy commented on src/nupic/engine/Link.cpp:
```diff
@@ -396,21 +556,49 @@ void Link::write(LinkProto::Builder& proto) const
 {
   std::ostream& operator<<(std::ostream& f, const Link& link)
   {
+    // TODO Shouldn't this be implemented in terms of Link::toString? (assuming
```
I think `toString` is more for debugging and not suitable for serialization. Plus it allocates a new string rather than add onto an existing stream.

vitaly-krugl replied:
@scottpurdy, I was thinking that `operator<<` could call `toString` to generate the content, then dump the string to the stream. If `toString` is for debugging, why does it render only a subset of the data versus `operator<<`? If `toString` rendered the same info that `operator<<` does right now, then the logic could be shared between the two via `toString`. Wouldn't it be nice for consistency?

scottpurdy replied:
The serialization code is a very specific thing with known requirements and backwards compatibility issues. So I wouldn't want it to delegate work to a `toString` function that someone may change without understanding the serialization implications. But it's also for the old serialization so I wouldn't spend too much time on it.

vitaly-krugl replied:
Fair enough, I will remove the TODO comment

vitaly-krugl commented:
@scottpurdy, I found/fixed the cause of SEGFAULT in TestNode. Also, the ArrayProto copy logic is in separate utils module now.

vitaly-krugl commented on src/nupic/engine/Input.cpp:
```diff
@@ -53,9 +53,9 @@ Input::~Input()
+
 void
-Input::addLink(const std::string& linkType, const std::string& linkParams,
-               Output* srcOutput, const size_t propagationDelay)
```
Replaced by orthogonal Link constructor plus the new addLink below that adds a Link instance to the Input/Output

vitaly-krugl commented on src/nupic/engine/TestNode.cpp:
```diff
@@ -116,7 +117,9 @@ namespace nupic
   TestNode::TestNode(AnyPointer::Reader& proto, Region* region) :
-    RegionImpl(region)
+    RegionImpl(region),
+    computeCallback_(nullptr)
```
The missing `computeCallback_(nullptr)` was resulting in sporadic SEGFAULT when running the new link serialization test.

vitaly-krugl commented:
cc @scottpurdy 
Please review;

scottpurdy commented on src/nupic/proto/LinkProto.capnp:
```diff
@@ -10,4 +12,8 @@ struct LinkProto {
   destInput @5 :Text;
+
+  outputElementType @6 :Int32; # NTA_BasicType: type of output elements in output array
+  outputElementCount @7 :UInt32; # number of output elements in output array
```
Is `outputElementCount` necessary? I believe that when you deserialize `delayedOutputs` you can check how many elements there are.

vitaly-krugl replied:
Thanks @scottpurdy, good catch.

vitaly-krugl replied:
Done via commit https://github.com/numenta/nupic.core/pull/1182/commits/0f57bb371a5b9800f1a4010911de6af3859ae729

scottpurdy commented on src/nupic/proto/ArrayProto.capnp:
```diff
@@ -0,0 +1,16 @@
+# Represents an Output Array
+struct ArrayProto {
+  union {
+    byteArray   @0 :List(UInt8);   # NTA_BasicType_Byte ZZZ or is it Int8, or does it matter?
```
Need to address your comment here

vitaly-krugl replied:
done

scottpurdy commented on src/nupic/proto/LinkProto.capnp:
```diff
@@ -10,4 +12,7 @@ struct LinkProto {
   destRegion @4 :Text;
   destInput @5 :Text;
+
+  outputElementType @6 :Int32; # NTA_BasicType: type of output elements in output array
```
Can this be determined by checking which field in `delayedOutputs` instances of `ArrayProto` is set?

vitaly-krugl commented:
@scottpurdy, I added Region::purgeInputLinkBufferHeads to engine_internal binding and updated the code per your feedback. Please review the latest changes. Thx.

scottpurdy commented on src/nupic/bindings/engine_internal.i:
```diff
@@ -46,6 +46,8 @@
 
+#include <nupic/engine/Input.hpp>
+#include <nupic/engine/Link.hpp>
 
```
No need for blank line to separate your includes from the rest of them

vitaly-krugl replied:
I followed what appears to be a precedent in this file, whereby the `nupic/utils/*` includes were separated by a blank line from the types/ntypes includes.

scottpurdy replied:
👍

scottpurdy commented on src/nupic/bindings/engine_internal.i:
```diff
@@ -46,6 +46,8 @@
  * ----------------------------------------------------------------------
```
Update both copyright headers to current year (2013-2017)

scottpurdy commented on src/nupic/bindings/engine_internal.i:
```diff
@@ -188,6 +192,18 @@
+  // Purge heads of the region's input link buffers
+  void purgeInputLinkBufferHeads()
+  {
+    for (auto & input : self->getInputs())
```
No need to change but just a note that if we move the circular buffer into the output in the future then this will need to be changed so that we don't purge multiple times in the case of two inputs linked to the same output.

vitaly-krugl replied:
That makes sense. Should I add a note about this somewhere, or just make a mental note of it?

scottpurdy replied:
Mental note is fine since this solution is fine in current implementation

scottpurdy commented:
👍 to merge when green
